### PR TITLE
v2: do not return objects from sign() functions anymore. Add setAddressIndex()

### PR DESCRIPTION
**Pull request by andreibancioiu**
- Do not return objects from `sign()` functions anymore
 - Add `setAddressIndex(addressIndex: number)`
 - Bit of refactoring
 - Updated examples

### CODE REVIEW

**Review by bogdan-rosianu — DISMISSED** [review]

File: examples/index.html [hunk 1]
```diff
@@ -3,21 +3,74 @@
...
+        async function init() {
+            await provider.init();
+            console.log("Initialized.");
+        }
+
+        async function login() {
+            var addressIndex = parseInt(document.getElementsByName("addressIndexForLogin")[0].value);
+            console.log("Login with addressIndex", addressIndex);
+            await provider.login({ addressIndex: addressIndex });
+            await dumpAddresses();
```
Comment (StanislavSava) [inline examples/index.html]: dumpAddresses sounds a bit like dropAddresses, can be confusing a bit, maybe "logAddresses"?

File: examples/index.html [hunk 2]
```diff
@@ -3,21 +3,74 @@
...
+
+        async function dumpAddresses() {
+            let addresses = await provider.getAccounts();
+            let currentAddress = await provider.getAddress();
+
+            console.log("Addresses:", addresses);
+            console.log("Current address:", currentAddress);
+        }
+
+        async function setAddress() {
```
Comment (StanislavSava) [inline examples/index.html]: sounds more like "setAddressIndex"

File: src/hwProvider.ts [hunk 3]
```diff
@@ -73,22 +74,27 @@ export class HWProvider implements IHWProvider {
      */
     async login(options?: { addressIndex?: number }): Promise<string> {
```
Comment (StanislavSava) [inline src/hwProvider.ts]: you could use addressIndex = 0: number

File: src/hwProvider.ts [hunk 4]
```diff
@@ -73,22 +74,27 @@ export class HWProvider implements IHWProvider {
...
         if (!this.hwApp) {
-            throw new Error("HWApp not initialised, call init() first");
+            throw new ErrNotInitialized();
         }
 
-        if(options && options.addressIndex) {
-            this.addressIndex = options.addressIndex;
-        }
+        let addressIndex = options?.addressIndex || 0;
```
Comment (StanislavSava) [inline src/hwProvider.ts]: and drop this line

File: src/hwProvider.ts [hunk 5]
```diff
@@ -73,22 +74,27 @@ export class HWProvider implements IHWProvider {
...
+        await this.setAddressIndex(addressIndex);
+        const { address } = await this.hwApp.getAddress(0, addressIndex, true);
```
Comment (StanislavSava) [inline src/hwProvider.ts]: Should we maybe handle for errors here and throw custom error messages to the user?

File: src/hwProvider.ts [hunk 6]
```diff
@@ -73,22 +74,27 @@ export class HWProvider implements IHWProvider {
...
+        return address;
+    }
 
-        await this.hwApp.setAddress(0, this.addressIndex);
-        const {address} = await this.hwApp.getAddress(0, this.addressIndex, true);
+    async setAddressIndex(addressIndex: number): Promise<void> {
```
Comment (StanislavSava) [inline src/hwProvider.ts:85]: YES!

File: src/hwProvider.ts [hunk 7]
```diff
@@ -140,18 +151,18 @@ export class HWProvider implements IHWProvider {
         return transaction;
     }
 
-    async signTransactions(transactions: ITransaction[]): Promise<ITransaction[]> {
-        let retTx: ITransaction[] = [];
+    async signTransactions<T extends ITransaction>(transactions: T[]): Promise<T[]> {
+        let result: T[] = [];
         for (let tx of transactions) {
```
Comment (StanislavSava) [inline src/hwProvider.ts:152]: I think a map would be cleaner here, with an async function inside

**Review by StanislavSava — COMMENTED** [review]
A couple of style comments

File: examples/index.html (hunk 1 above)
Comment (andreibancioiu) [reply examples/index.html]: Fixed (renamed).

File: examples/index.html (hunk 2 above)
Comment (andreibancioiu) [reply examples/index.html]: Fixed (renamed).

File: src/hwProvider.ts (hunk 3 above)
Comment (andreibancioiu) [reply src/hwProvider.ts]: Fixed, added default.

File: src/hwProvider.ts (hunk 4 above)
Comment (andreibancioiu) [reply src/hwProvider.ts]: Removed.

File: src/hwProvider.ts (hunk 5 above)
Comment (andreibancioiu) [reply src/hwProvider.ts]: Indeed, some error catching and wrapping would be good. Is it OK to postpone this to a separate PR? Currently, it's same as _before_.

File: src/hwProvider.ts (hunk 7 above)
Comment (andreibancioiu) [reply src/hwProvider.ts:152]: Simplified (returning the same array as the input).

**Review by StanislavSava — APPROVED** [review]

**Review by bogdan-rosianu — APPROVED** [review]

**Pull request merged**